rust/ql/lib/codeql/rust/internal/Type.qll

```diff
@@ -16,10 +16,13 @@ newtype TType =
   TArrayType() or // todo: add size?
   TRefType() or // todo: add mut?
   TImplTraitType(ImplTraitTypeRepr impl) or
+  TSliceType() or
   TTypeParamTypeParameter(TypeParam t) or
   TAssociatedTypeTypeParameter(TypeAlias t) { any(TraitItemNode trait).getAnAssocItem() = t } or
+  TArrayTypeParameter() or
   TRefTypeParameter() or
-  TSelfTypeParameter(Trait t)
+  TSelfTypeParameter(Trait t) or
+  TSliceTypeParameter()
```

Comment:
Out of interest, how are these different?  i.e. why is just one `TypeParameter()` type insufficient?

Reply:
We model slice types as if there was a special `struct Slice<T> = ...` type, and `TSliceTypeParameter` corresponds to the `T`. So e.g. `[i64]` translates to `Slice<i64>`.